### PR TITLE
Add optional Parquet serializer to cube interface

### PR DESCRIPTION
**Pull request by stephan-hesselmann-by**
This allows to specify the compression algorithm and a chunk size to
generate row groups.

The Parquet serializer parameter is added the following cube operations:
  * `build_cube`
  * `append_cube`
  * `extend_cube`
  * `update_cube`

Specifying different Parquet serializers is back- and forwards compatible.
Row groups are kept for partitions that are not touched during cube operations.
If partitions are replaced or newly created, they are rewritten by the
specified Parquet serializer.

# Description:

Briefly describe the change of behavior

- [ ] Closes #xxxx
- [ ] Changelog entry

### CODE REVIEW

File: kartothek/io/testing/build_cube.py [hunk 1]
```diff
@@ -269,6 +273,52 @@ def test_parquet(driver, function_store):
         pdt.assert_frame_equal(df_actual.reset_index(drop=True), df_expected)
 
 
+@pytest.mark.filterwarnings("ignore::UnicodeWarning")
+@pytest.mark.parametrize("chunk_size", [None, 1, 2, 3])
```
Comment (fjetter) [inline kartothek/io/testing/build_cube.py:277]: `[None, >0]` should be sufficient. No need to have multiple integers in here

File: kartothek/io/testing/build_cube.py [hunk 2]
```diff
@@ -269,6 +273,52 @@ def test_parquet(driver, function_store):
         pdt.assert_frame_equal(df_actual.reset_index(drop=True), df_expected)
 
 
+@pytest.mark.filterwarnings("ignore::UnicodeWarning")
```
Comment (fjetter) [inline kartothek/io/testing/build_cube.py:276]: Where is this coming from?

File: kartothek/io/testing/extend_cube.py [hunk 3]
```diff
@@ -79,6 +84,86 @@ def test_simple(driver, function_store, existing_cube):
     assert set(ds.table_meta) == {SINGLE_TABLE}
 
 
+@pytest.mark.parametrize("chunk_size", [None, 1, 2, 3])
```
Comment (fjetter) [inline kartothek/io/testing/extend_cube.py:87]: same here, please reduce the multiplicity

File: kartothek/io/testing/extend_cube.py [hunk 4]
```diff
@@ -79,6 +84,86 @@ def test_simple(driver, function_store, existing_cube):
...
+
+    part_num_rows = {0: 1, 1: 3}
+    part_chunk_size = {0: None, 1: None}
+
+    assert len(dataset.partitions) == 2
+    assert_num_row_groups(function_store(), dataset, part_num_rows, part_chunk_size)
+
+
+@pytest.mark.parametrize(
+    "compression_build", ["NONE", "SNAPPY", "GZIP", "BROTLI", "LZ4", "ZSTD"]
```
Comment (fjetter) [inline kartothek/io/testing/extend_cube.py]: same here, please reduce the multiplicity

File: kartothek/serialization/_parquet.py [hunk 5]
```diff
@@ -69,8 +69,22 @@ class ParquetSerializer(DataFrameSerializer):
     _PARQUET_VERSION = "2.0"
     type_stable = True
 
-    def __init__(self, compression="SNAPPY", chunk_size=None):
+    def __init__(
+        self, compression: str = "SNAPPY", chunk_size: Optional[int] = None
+    ) -> None:
+        if compression not in ["NONE", "SNAPPY", "GZIP", "BROTLI", "LZ4", "ZSTD"]:
```
Comment (fjetter) [inline kartothek/serialization/_parquet.py:75]: I would assume this list is exported by arrow. Can you please check and reuse an arrow interface? Which algo is supported is 100% dictated by arrow and depends even on how it was built, therefore, I would prefer a more dynamic approach, if any

**Review by fjetter — CHANGES_REQUESTED** [review]

File: kartothek/io/testing/append_cube.py [hunk 6]
```diff
@@ -98,6 +103,117 @@ def test_append_partitions(driver, function_store, existing_cube):
...
+
+    part_num_rows = {0: 2, 1: 2, 2: 1, 3: 3}
+    part_chunk_size = {0: None, 1: None, 2: None, 3: None}
+
+    assert len(dataset.partitions) == 4
+    assert_num_row_groups(function_store(), dataset, part_num_rows, part_chunk_size)
+
+
+@pytest.mark.parametrize(
+    "compression_build", ["NONE", "SNAPPY", "GZIP", "BROTLI", "LZ4", "ZSTD"]
```
[the remark this answers is not in the document]
Comment (stephan-hesselmann-by) [reply kartothek/io/testing/append_cube.py]: Would you agree that the test itself makes sense if the parameters are removed/hardcoded to e.g. SNAPPY and GZIP?

File: kartothek/io/testing/build_cube.py (hunk 1 above)
Comment (stephan-hesselmann-by) [reply kartothek/io/testing/build_cube.py:277]: Note that these do test different edge cases (e.g. whether number of rows is divisible by chunk size). But I agree this is probably the wrong abstraction level to test this.

File: kartothek/io/testing/append_cube.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fjetter) [reply kartothek/io/testing/append_cube.py]: from my POV what you are trying to test is "what happens if I create + update with differently configured serialiser instances." From that angle, I see value and it is good to have this explicit.
however, this is technically already done with the test on different chunk sizes above so there is some redundancy in this test, already.
If we keep one combination hard coded, with a comment next to it explaining why it is not parametrized, I'd be fine with keeping one compression test if you feel it adds value

File: kartothek/serialization/_parquet.py (hunk 5 above)
Comment (stephan-hesselmann-by) [reply kartothek/serialization/_parquet.py:75]: I didn't find an exported list by arrow on first glance, perhaps because it is indeed based on the arrow build. I will just remove this check

File: kartothek/io/testing/build_cube.py (hunk 1 above)
Comment (fjetter) [reply kartothek/io/testing/build_cube.py:277]: indeed, the test makes sense but not on cube level. the only thing we actually want to test on cube level is "did I actually use the provided serializer or not".

You could even take this a level further by implementing a mock serializer and we merely verify if it was called. This way we'd not do any IO in the test. After all, we rely on the serializer to be implemented correctly so we don't need to test all peculiarities of it. We might finally have fast tests this way :D

Note: Completely fine if you don't go down that route but eventually I would hope we can test kartothek on this level instead of always performing full end-to-end tests

**Review by fjetter — APPROVED** [review]
LGTM. Thanks @stephan-hesselmann-by!

File: kartothek/io/dask/bag_cube.py [hunk 7]
```diff
@@ -39,35 +44,38 @@
...
-    cube,
-    store,
-    ktk_cube_dataset_ids=None,
-    metadata=None,
-    overwrite=False,
-    partition_on=None,
-):
+    data: db.Bag,
+    cube: Cube,
+    store: Callable[[], KeyValueStore],
```
Comment (fjetter) [inline kartothek/io/dask/bag_cube.py:49]: FYI We have a type alias `StoreFactory` for this in `kartothek.core.typing`

Comment (stephan-hesselmann-by) [reply kartothek/io/dask/bag_cube.py:49]: Thanks for pointing this out. The type hints are overall quite confusing because sometimes store is a callable and sometimes not. I have simply followed the convention from the previous doc strings.

Comment (fjetter) [reply kartothek/io/dask/bag_cube.py:49]: Indeed, this is very confusing. With https://github.com/JDASoftwareGroup/kartothek/pull/345 I tried to change the behaviour such that user facing code just doesn't care but this was not done for the cubes and everything is overall pretty messy

**Comment by stephan-hesselmann-by** [comment]
Rebased after https://github.com/JDASoftwareGroup/kartothek/pull/398 was merged
